# Introduction to Data Analysis with Python

Estimated time needed: **10** minutes

## Objectives

- Acquire data in various ways
- Obtain insights from data with Pandas library

## Table of Contents

- [Heading 1](#heading-1)

---

## Video Lessons

- <img src="https://user-images.githubusercontent.com/95575665/168158125-0d826c1a-ffea-4ea9-a2bc-f3b0f4f0b99a.png" alt="youtube-logo.png" width="15" height="10"> [Introduction to Data Analysis with Python](https://www.youtube.com/watch?v=egFbY9so2XM)
- <img src="https://user-images.githubusercontent.com/95575665/168158125-0d826c1a-ffea-4ea9-a2bc-f3b0f4f0b99a.png" alt="youtube-logo.png" width="15" height="10"> [The Problem](https://www.youtube.com/watch?v=mMmMZIMPM34)
- <img src="https://user-images.githubusercontent.com/95575665/168158125-0d826c1a-ffea-4ea9-a2bc-f3b0f4f0b99a.png" alt="youtube-logo.png" width="15" height="10"> [Understanding the Data](https://www.youtube.com/watch?v=JGWKVhKb-bo)
- <img src="https://user-images.githubusercontent.com/95575665/168158125-0d826c1a-ffea-4ea9-a2bc-f3b0f4f0b99a.png" alt="youtube-logo.png" width="15" height="10"> [Python Packages for Data Science](https://www.youtube.com/watch?v=rTNd5SlYcDg)
- <img src="https://user-images.githubusercontent.com/95575665/168158125-0d826c1a-ffea-4ea9-a2bc-f3b0f4f0b99a.png" alt="youtube-logo.png" width="15" height="10"> [Importing and Exporting Data in Python](https://www.youtube.com/watch?v=aT8H_XcdirY)
- <img src="https://user-images.githubusercontent.com/95575665/168158125-0d826c1a-ffea-4ea9-a2bc-f3b0f4f0b99a.png" alt="youtube-logo.png" width="15" height="10"> [Getting Started Analyzing Data in Python](https://www.youtube.com/watch?v=RiY6Zce_bbg)

---

## Data Acquisition

In this section, we will learn how to load a dataset into our Jupyter Notebook. Datasets can be stored on our local machine or on the internet. There are various formats for a dataset: `.csv`, `.json`, `.xlsx`, etc. In our case, the automobile dataset is an online source and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.

- Data source: [https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data](https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data)
- Data type: CSV

The Pandas library is a useful tool that enables us to read various datasets into a dataframe.

In [1]:
## Install a specific version of libraries
# !mamba install pandas=1.3.3 -y
# !mamba install numpy=1.21.2 -y

`mamba` is a reimplementation of the Conda package manager in C++, designed to be fast. To install Mamba, go to [Installation](https://mamba.readthedocs.io/en/latest/installation.html) of [Mamba documentation](https://mamba.readthedocs.io/en/latest/index.html).

In [2]:
import pandas as pd
import numpy as np
import requests
import io

## Read Data

We use `read_csv()` function to read the CSV file. We put the file path in quotation marks in the brackets so that Pandas will read the file into a DataFrame from that address. The file path can be either an URL or a local file address.

Since the data does not include headers, we should add an argument `headers=None` inside the `read_csv()` method so that pandas will not automatically set the first row as a header.

In [3]:
## Data file in .data format stored on the internet
file_path = "https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data"
## Data file in .data format stored on local machine
# file_path = "imports-85.data"
## Data file in .csv format stored on the internet
# file_path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"

## The get() method sends a GET request to the specified URL and returns a response object.
## "content" returns the content of the response, in bytes.
# content = requests.get(file_path).content
# c = pd.read_csv(io.StringIO(content.decode('utf-8')))
# print(c)

# read_file = pd.read_csv(file_path)
# read_file.to_csv(r'data.csv', index=None)
# print(read_file)

## Read the data file stored online by the URL provided
df = pd.read_csv(file_path, header=None)

After reading the dataset, we can use the `head(N)` method to check the top N rows of the DataFrame, where N is an integer. Contrary to `head()`, `tail(N)` shows the bottom N rows of the DataFrame.

In [4]:
## Show the 1st 5 rows of the dataset 
df.head()

0    1            2    3    4     5            6    7      8     9   ...  \
0   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
1   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
2   1    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...   
3   2  164         audi  gas  std  four        sedan  fwd  front  99.8  ...   
4   2  164         audi  gas  std  four        sedan  4wd  front  99.4  ...   

    16    17    18    19    20   21    22  23  24     25  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  

[5 rows x 26 columns]

### Add Headers

Take a look at the dataset. Pandas automatically sets the header with an integer starting from 0. To better describe our data, we can introduce a header. This information is available at [Attribute Information](https://archive.ics.uci.edu/ml/datasets/Automobile). We have to add headers manually. First, we create a list `headers` that include all column names in order. Then, we use `df.columns=headers` to replace the headers with the list we created.

In [5]:
## Create headers list
headers = ["symboling", "normalized-losses", "make", "fuel-type", "aspiration", "num-of-doors", "body-style", "drive-wheels", "engine-location", "wheel-base", "length", "width", "height", "curb-weight", "engine-type", "num-of-cylinders", "engine-size", "fuel-system", "bore", "stroke", "compression-ratio", "horsepower", "peak-rpm", "city-mpg", "highway-mpg", "price"]
print(headers)

['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


In [6]:
## Replace headers 
df.columns = headers
## Recheck DataFrame
df.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2                 ?         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1                 ?         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0                 ?         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   
5        sedan          fwd           front        99.8  ...          136   
6        sedan          fwd           front       105.8  ...          136   
7        wagon          fwd           front       105.8  ...          136   
8        sedan          fwd           front       105.8  ...          131   
9    hatchback          4wd           front        99.5  ...          131   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   
5         mpfi  3.19    3.40               8.5        110      5500       19   
6         mpfi  3.19    3.40               8.5        110      5500       19   
7         mpfi  3.19    3.40               8.5        110      5500       19   
8         mpfi  3.13    3.40               8.3        140      5500       17   
9         mpfi  3.13    3.40               7.0        160      5500       16   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  
5          25  15250  
6          25  17710  
7          25  18920  
8          20  23875  
9          22      ?  

[10 rows x 26 columns]

We need to replace the "?" symbol with NaN so the dropna() can remove the missing values.

In [7]:
df_nan = df.replace('?', np.NaN)

In [8]:
## Drop missing values along the column "price"
df = df_nan.dropna(subset=["price"], axis=0)
df.head(20)

symboling normalized-losses         make fuel-type aspiration  \
0           3               NaN  alfa-romero       gas        std   
1           3               NaN  alfa-romero       gas        std   
2           1               NaN  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2               NaN         audi       gas        std   
6           1               158         audi       gas        std   
7           1               NaN         audi       gas        std   
8           1               158         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1               NaN          bmw       gas        std   
15          0               NaN          bmw       gas        std   
16          0               NaN          bmw       gas        std   
17          0               NaN          bmw       gas        std   
18          2               121    chevrolet       gas        std   
19          1                98    chevrolet       gas        std   
20          0                81    chevrolet       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   
15         four        sedan          rwd           front       103.5  ...   
16          two        sedan          rwd           front       103.5  ...   
17         four        sedan          rwd           front       110.0  ...   
18          two    hatchback          fwd           front        88.4  ...   
19          two    hatchback          fwd           front        94.5  ...   
20         four        sedan          fwd           front        94.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
10          108         mpfi  3.50    2.80               8.8        101   
11          108         mpfi  3.50    2.80               8.8        101   
12          164

Now, we have successfully read the raw dataset and added the correct headers into the DataFrame.

## Save Dataset

Correspondingly, Pandas enables us to save the dataset to CSV by using the `to_csv()` method. For example, we may use the syntax below, where `index=False` means the row names will not be written, to save the DataFrame to our local machine.

In [9]:
## Save the DataFrame as CSV file to local machine
df.to_csv("automobile.csv", index=False)

We can also read and save other data file formats. These functions are listed below.

### Read and Save Other Data Formats

| Data Format | File Extension | Read Function     | Save Function   |
| :---------: | :------------: | :---------------- | :-------------- |
| CSV         | .csv           | `pd.read_csv()`   | `df.to_csv()`   |
| JSON        | .json          | `pd.read_json()`  | `df.to_json()`  |
| Excel       | .xlsx          | `pd.read_excel()` | `df.to_excel()` |
| SQL         | .sql           | `pd.read_sql()`   | `df.to_sql()`   |
| HDF         | .hdf           | `pd.read_hdf()`   | `df.to_hdf()`   |

## Insight of Dataset

After reading data into Pandas DataFrame, it is time for us to explore the dataset. There are several ways to obtain essential insights of the data to help us better understand our dataset.

### Data Types

Data has a variety of types. The main types stored in Pandas dataframes are `object`, `int64`, `float64`, `bool` and `datetime64`. In order to better learn about each attribute, it is always good for us to know the data type of each column. We can use `dtypes` to check this. A series with the data type of each column will be returned.

In [10]:
## Check the data type of each column
print(df.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


As shown above, the data type of "symboling", "normalized-losses" and "wheel-base" are `int64`, `object` and `float64`, respectively. These data types can be changed.

### Describe

To get a statistical summary of each column, e.g., count, column mean value, column standard deviation, etc., we can use the `describe()` method. This method will provide various summary statistics, excluding `NaN` (Not a Number) values.

In [11]:
## Generate descriptive statistics
df.describe()

symboling  wheel-base      length       width      height  \
count  201.000000  201.000000  201.000000  201.000000  201.000000   
mean     0.840796   98.797015  174.200995   65.889055   53.766667   
std      1.254802    6.066366   12.322175    2.101471    2.447822   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.800000   64.100000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.500000   66.600000   55.500000   
max      3.000000  120.900000  208.100000   72.000000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   201.000000   201.000000         201.000000  201.000000   201.000000  
mean   2555.666667   126.875622          10.164279   25.179104    30.686567  
std     517.296727    41.546834           4.004965    6.423220     6.815150  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2169.000000    98.000000           8.600000   19.000000    25.000000  
50%    2414.000000   120.000000           9.000000   24.000000    30.000000  
75%    2926.000000   141.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

This shows the statistical summary of all numeric-typed columns. For example, the attribute "symboling" has 205 counts, the mean value of this column is 0.83, the standard deviation is 1.25, the minimum value is -2, 25th percentile is 0, 50th percentile is 1, 75th percentile is 2, and the maximum value is 3. However, what if we would also like to check all the columns including those that are of type object? We can add an argument `include="all"`.

In [12]:
## Describe all the columns in the DataFrame 
df.describe(include="all")

symboling normalized-losses    make fuel-type aspiration  \
count   201.000000               164     201       201        201   
unique         NaN                51      22         2          2   
top            NaN               161  toyota       gas        std   
freq           NaN                11      32       181        165   
mean      0.840796               NaN     NaN       NaN        NaN   
std       1.254802               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           199        201          201             201  201.000000  ...   
unique            2          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            113         94          118             198         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.797015  ...   
std             NaN        NaN          NaN             NaN    6.066366  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    201.000000          201   197     197        201.000000        199   
unique          NaN            8    38      36               NaN         58   
top             NaN         mpfi  3.62    3.40               NaN         68   
freq            NaN           92    23      19               NaN         19   
mean     126.875622          NaN   NaN     NaN         10.164279        NaN   
std       41.546834          NaN   NaN     NaN          4.004965        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       98.000000          NaN   NaN     NaN          8.600000        NaN   
50%      120.000000          NaN   NaN     NaN          9.000000        NaN   
75%      141.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        199  201.000000  201.000000   201  
unique        22         NaN         NaN   186  
top         5500         NaN         NaN  8921  
freq          36         NaN         NaN     2  
mean         NaN   25.179104   30.686567   NaN  
std          NaN    6.423220    6.815150   NaN  
min          NaN   13.000000   16.000000   NaN  
25%          NaN   19.000000   25.000000   NaN  
50%          NaN   24.000000   30.000000   NaN  
75%          NaN   30.000000   34.000000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

Now it provides the statistical summary of all the columns, including object-typed attributes. We can now see how many unique values are there, which one is the top value and the frequency of the top value in the object-typed columns. Some values in the table above show as "NaN". This is because those numbers are not available regarding a particular column type.

### Info

Another method we can use to check our dataset is `info()`, which provides a concise summary of a DataFrame. This method prints information about a DataFrame, including the index dtype and columns, non-null values and memory usage.

In [13]:
## Get a concise summary of a DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       199 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non

---

## Exercises

1. Check the bottom 10 rows of the DataFrame `df`.

```python
df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv", header=None)
```

In [14]:
df.tail(10)

symboling normalized-losses   make fuel-type aspiration num-of-doors  \
195         -1                74  volvo       gas        std         four   
196         -2               103  volvo       gas        std         four   
197         -1                74  volvo       gas        std         four   
198         -2               103  volvo       gas      turbo         four   
199         -1                74  volvo       gas      turbo         four   
200         -1                95  volvo       gas        std         four   
201         -1                95  volvo       gas      turbo         four   
202         -1                95  volvo       gas        std         four   
203         -1                95  volvo    diesel      turbo         four   
204         -1                95  volvo       gas      turbo         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
195      wagon          rwd           front       104.3  ...          141   
196      sedan          rwd           front       104.3  ...          141   
197      wagon          rwd           front       104.3  ...          141   
198      sedan          rwd           front       104.3  ...          130   
199      wagon          rwd           front       104.3  ...          130   
200      sedan          rwd           front       109.1  ...          141   
201      sedan          rwd           front       109.1  ...          141   
202      sedan          rwd           front       109.1  ...          173   
203      sedan          rwd           front       109.1  ...          145   
204      sedan          rwd           front       109.1  ...          141   

     fuel-system  bore  stroke compression-ratio horsepower  peak-rpm  \
195         mpfi  3.78    3.15               9.5        114      5400   
196         mpfi  3.78    3.15               9.5        114      5400   
197         mpfi  3.78    3.15               9.5        114      5400   
198         mpfi  3.62    3.15               7.5        162      5100   
199         mpfi  3.62    3.15               7.5        162      5100   
200         mpfi  3.78    3.15               9.5        114      5400   
201         mpfi  3.78    3.15               8.7        160      5300   
202         mpfi  3.58    2.87               8.8        134      5500   
203          idi  3.01    3.40              23.0        106      4800   
204         mpfi  3.78    3.15               9.5        114      5400   

    city-mpg highway-mpg  price  
195       23          28  13415  
196       24          28  15985  
197       24          28  16515  
198       17          22  18420  
199       17          22  18950  
200       23          28  16845  
201       19          25  19045  
202       18          23  21485  
203       26          27  22470  
204       19          25  22625  

[10 rows x 26 columns]

<details><summary>Click here for the solution</summary>

```python
df.tail(10)
```

</details>

2. Find the name of the columns of the DataFrame.

In [15]:
df.columns

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')

<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>

3. We can use the following syntax to get the statistics of particular columns, where we indicate the name of the selected columns in square brackets (`[]`).
```
df[['column_name_1', 'column_name_2', 'column_name_3', ...]].describe()
```
Apply the method `describe()` to the columns "length" and "compression-ratio".

In [16]:
df[["length", "compression-ratio"]].describe()

length  compression-ratio
count  201.000000         201.000000
mean   174.200995          10.164279
std     12.322175           4.004965
min    141.100000           7.000000
25%    166.800000           8.600000
50%    173.200000           9.000000
75%    183.500000           9.400000
max    208.100000          23.000000

<details><summary>Click here for the solution</summary>

```python
df[['length', 'compression-ratio']].describe()
```

</details>

---

Author(s):

- [Joseph Santarcangelo](https://www.linkedin.com/in/joseph-s-50398b136)

Contributor(s):

- [Mahdi Noorian PhD](https://www.linkedin.com/in/mahdi-noorian-58219234)
- Bahare Talayian
- Eric Xiao
- Steven Dong
- Parizad
- Hima Vasudevan
- [Fiorella Wenver](https://www.linkedin.com/in/fiorellawever)
- [Yi Yao](https://www.linkedin.com/in/yi-leng-yao-84451275)